src/enemy/FireyPetey.cpp

```diff
@@ -1,456 +1,746 @@
+                target->mWaterEmitInfo->mPos.set(jointTranslation);
+                gpModelWaterManager->emitRequest(*target->mWaterEmitInfo);
+            }
+
+        }
+        if (frameCtrl->mCurFrame > 35.0f) {
+            target->_10 = 1;
+        }
+    }
+    return false;
+}
+
+bool TNerveFPTumble::execute(TSpineBase<TLiveActor> *spine) const {
+    TFireyPetey *target = reinterpret_cast<TFireyPetey *>(spine->mTarget);
+    TBossPakkunParams *params = reinterpret_cast<TBossPakkunParams *>(target->getSaveParam());
+    if (spine->mNerveTimer == 0) {
+        target->changeBck(0x6);
+        target->_02 = 1;
+    }
+
+    if (spine->mNerveTimer < params->mSLTumbleTime.get()) {
+        return false;
+    }
+    spine->pushNerve(&tumbleOut);
+    return true;
+}
+
+
+TFireyPetey::TFireyPetey(const char *test)
+    : TBossPakkun(test), mMActorKeeperSecondary(nullptr), mKukkuBall{}, mFire{} {
+    mNumTornados = MAX_TORNADOS;
+}
+
+void TFireyPetey::init(TLiveManager *liveManager) {
+    TBossPakkun::init(liveManager);
+
+    mTornado = new TFPTornado(this, "<TFPTornado>\n");
+    mNavel = new TFPNavel(this, "<TFPNavel>\n");
+    mHeadHit = new TFPHeadHit(this, "<TFPHeadHit>\n");
+
+    JDrama::TNameRefGen *instance = JDrama::TNameRefGen::getInstance();
+    TNameRef *root = instance->getRootNameRef();
+
+    u16 keyCode = calcKeyCode(gUnkManagerName);
+
+    TIdxGroupObj* group = static_cast<TIdxGroupObj *>(root->searchF(keyCode, gUnkManagerName));
+
+    group->mViewObjList.insert(group->mViewObjList.end(), mTornado);
+    group->mViewObjList.insert(group->mViewObjList.end(), mNavel);
+    group->mViewObjList.insert(group->mViewObjList.end(), mHeadHit);
+
+    mSpineBase->setNerve(&sleep);
+
+    mMActorKeeperSecondary = new TMActorKeeper(mLiveManager, NUM_GOOP_DROPS);
+
+    for (int i = 0; i < NUM_GOOP_DROPS; i++) {
+        mKukkuBall[i] = new TKukkuBall("lavaDrop\n");
+        mKukkuBall[i]->mActorData = mMActorKeeperSecondary->createMActor("torifun.bmd", 3);
+        mKukkuBall[i]->init();
+    }
+
+    for (int i = 0; i < 10; i++) {
+        mFire[i] = new TFPFire("fire\n");
+        group->mViewObjList.insert(group->mViewObjList.end(), mFire[i]);
+    }
+}
+
+void TFireyPetey::perform(u32 flags, JDrama::TGraphics *graphics) {
+    TBossPakkun::perform(flags, graphics);
+    for (int i = 0; i < NUM_GOOP_DROPS; i++) {
+        if (mKukkuBall[i] != nullptr) {
+            mKukkuBall[i]->perform(flags, graphics);
+        }
+    }
+    for (int i = 0; i < 10; i++) {
+        if (mFire[i] != nullptr) {
+            mFire[i]->perform(flags, graphics);
+        }
+    }
+}
+
+bool TFireyPetey::receiveMessage(THitActor *reciever, u32 param2) {
+    if (reciever->mObjectID == 0x1000001 && mSpineBase->getLatestNerve() == &sleep) {
+        mSpineBase->setNerve(&breakSleep);
+        return false;
+    }
+    if (_02 == 3 && (reciever->mObjectID == 0x1000001) || (reciever->mObjectID == 0x100000d)) {
```

Comment:
Make defines for these magic numbers (the object IDs) to describe what they represent.